pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,479 @@
+    } else {
+      _pushLocations = null;
+    }
+
+    _logger.info("*********************************************************************");
+    _logger.info("enable.partitioning: {}", _enablePartitioning);
+    _logger.info("enable.sorting: {}", _enableSorting);
+    _logger.info("enable.resizing: {}", _enableResizing);
+    _logger.info("path.to.input: {}", _inputSegmentDir);
+    _logger.info("path.to.output: {}", _preprocessedOutputDir);
+    _logger.info("path.to.deps.jar: {}", _pathToDependencyJar);
+    _logger.info("push.locations: {}", _pushLocations);
+    _logger.info("*********************************************************************");
+  }
+
+  public void run() throws Exception {
+    if (!_enablePartitioning && !_enableSorting && !_enableResizing) {
+      _logger.info("Pre-computing job is disabled.");
+      return;
+    } else {
+      _logger.info("Starting {}", getClass().getSimpleName());
+    }
+
+    _fileSystem = FileSystem.get(_configuration);
+    final List<String> inputDataPath = Arrays.asList(_inputSegmentDir.split("\\s*,\\s*"));
+    Path outputDir = new Path(_preprocessedOutputDir);
+
+    if (_fileSystem.exists(outputDir)) {
+      _logger.warn("Found the output folder {}, deleting it", _preprocessedOutputDir);
+      _fileSystem.delete(outputDir, true);
+    }
+    JobPreparationHelper.setDirPermission(_fileSystem, outputDir, _defaultPermissionsMask);
+
+    _tableConfig = getTableConfig();
+
+    if (_enablePartitioning) {
+      fetchPartitioningConfig();
+      _logger.info("{}: {}", PARTITION_COLUMN, _partitionColumn);
+      _logger.info("{}: {}", NUMBER_OF_PARTITIONS, _numberOfPartitions);
+      _logger.info("{}: {}", PARTITION_FUNCTION, _partitionColumn);
+    }
+
+    if (_enableSorting) {
+      fetchSortingConfig();
+      _logger.info("{}: {}", SORTED_COLUMN, _sortedColumn);
+    }
+
+    if (_enableResizing) {
+      fetchResizingConfig();
+      _logger.info("minimum number of output files: {}", _numberOfOutputFiles);
+    }
+
+    _logger.info("Initializing a pre-computing job");
+    Job job = Job.getInstance(_configuration);
+
+    job.getConfiguration().set(JobContext.JOB_NAME, this.getClass().getName());
+    // Turn this on to always firstly use class paths that user specifies.
+    job.getConfiguration().set(MRJobConfig.MAPREDUCE_JOB_USER_CLASSPATH_FIRST, "true");
+    // Turn this off since we don't need an empty file in the output directory
+    job.getConfiguration().set(FileOutputCommitter.SUCCESSFUL_JOB_OUTPUT_DIR_MARKER, "false");
+
+    job.setJarByClass(SegmentPreprocessingJob.class);
+
+    String hadoopTokenFileLocation = System.getenv("HADOOP_TOKEN_FILE_LOCATION");
+    if (hadoopTokenFileLocation != null) {
+      job.getConfiguration().set("mapreduce.job.credentials.binary", hadoopTokenFileLocation);
+    }
+
+    // Schema configs.
+    Schema schema = getSchema(inputDataPath.get(0));
```

Comment:
are there any utils that can get this info? seems weird that we have to manually get 0th elem

Reply:
In Avro MR job, it requires to explicitly set the schema before running the job. Currently there's no way to get the avro schema beforehand though.